### ISSUE TITLE
Add support for Spark steps

### ISSUE THREAD

**Issue by rslifka** [rslifka/elasticity#85]

**Comment by robert2d** [comment]
I have managed to get this working today with the `ScriptStep` it was a bit annoying figuring out how to pass the arguments correctly(IE; all separate for flags and values). But its working fine.

What would be very useful is maybe a bit of documentation on how You submit a spark step. and more importantly as of ami_version 4.0 of EMR you can pass Applications to install of which Spark is one of them.. http://docs.aws.amazon.com/ElasticMapReduce/latest/API/API_RunJobFlow.html. At the moment I am using an older version and running some bootstrap steps and an additional step for adding spark history.

More than willing to put this all into a pully if that suits?

**Comment by rslifka** [comment]
Are you referring to the installation of Spark via `RunJobFlow` => `NewSupportedProducts`?  Or the submission of a Spark job (i.e. adding a step to the jobflow)?

There's a little behind the scenes magic happening with `HiveStep` and `PigStep` where if you've added one of those to your jobflow, it will automatically add a step to install them.  It would be nice to do the same thing for SparkStep.  If you've added a `SparkStep`, `JobFlow` detects this and makes sure to install the Spark application.

What do you think?

**Comment by robert2d** [comment]
Yea I like that, I was referring to `Applications.member.N` in that AWS API link. being able to create a jobflow and specify the packages required would be great. The reason for this is I start a cluster straight away then do a bunch of streaming jobs to process GZ to LZO(faster processing on Spark) so when I add the Spark job I want it to have spark already running so it just runs the Scala JAR right away.

I do however really like the Idea of adding a `SparkStep` which checks for the presence of spark in the JobFlow and adds it if required, that would be awesome!

So basically I think we should do both options? I can kick this off today hopefully with a proof of concept within out application for adding a `SparkStep` I'll bring that back into here for you too look over/in a pully.

**Comment by rslifka** [comment]
If you check out the `.aws_installation_step` stuff, it's actually quite similar.

There's a set of steps (Hive, Pig, etc.) that use steps to install software and a set of steps (Spark) that have required applications.  I think you can either repurpose what's there or do something in parallel (`.aws_required_application` perhaps).

Thanks in advance sir :smile:
